### ISSUE TITLE
How to use `prettier`'s `plugins` option?

### ISSUE THREAD

**Issue by srid** [numtide/treefmt-nix#112]
Follow-up to #103

When using the `prettier` formatter, how do we actually use the `plugins` option?

With

```nix
programs.prettier.settings = {
  plugins = [ "prettier-plugin-tailwindcss" ];
};
```

(To use https://tailwindcss.com/blog/automatic-class-sorting-with-prettier)

I get:

```
[INFO]: Error using formatter #prettier:
• [STDOUT]:

• [STDERR]:
[error] Cannot find package 'prettier-plugin-tailwindcss' imported from /Users/srid/code/nix-browser/noop.js
```

**Comment by gkze** [comment]
You'd need to install the NPM module yourself. Right now accessing/managing NPM modules from Nix/Nixpkgs isn't straightforward

**Comment by Mic92** [comment]
Does pluginSearchDirs has any effect at all? I couldn't find it in the upstream documentations. I packaged the plugin but it's not found by prettier:

```
    treefmt.programs.prettier.settings.plugins = [ "prettier-plugin-tailwindcss" ];
    treefmt.programs.prettier.settings.pluginSearchDirs = [
      "${self'.packages.prettier-plugin-tailwindcss}/lib"
    ];
```

**Comment by gkze** [comment]
Yeah it' not pretty (no pun intended lol) but that works... hopefully once NPM <=> Nix improves it will be more straightforward

**Comment by multivac61** [comment]
@Mic92 Is this still the recommended way? Looks like the absolutely plugin path hack has been removed from HEAD: https://git.clan.lol/clan/clan-core/src/branch/main/formatter.nix

**Comment by jukremer** [comment]
What is currently the best method to use a plugin that's in `nixpkgs`? 

```nix
prettier = {
  enable = true;
  settings = {
    plugins = [ "${pkgs.prettier-plugin-go-template}" ];
    overrides = {
      files = [ "*.html" ];
      options.parser = "go-template";
    };
  };
};
```

Would this work? Do I need to specify the path further? Do I have to add the package separately?

**Comment by Mic92** [comment]
> @Mic92 Is this still the recommended way? Looks like the absolutely plugin path hack has been removed from HEAD: https://git.clan.lol/clan/clan-core/src/branch/main/formatter.nix

I don't actually don't know, we currently don't need prettier.

**Comment by gkze** [comment]
When I got this working I installed the plugins via the devShell's packages and specified their names in the prettier config

**Comment by jukremer** [comment]
> Does pluginSearchDirs has any effect at all? I couldn't find it in the upstream documentations. I packaged the plugin with npm2nix but it's not found by prettier:
> 
> ```
>     treefmt.programs.prettier.settings.plugins = [ "prettier-plugin-tailwindcss" ];
>     treefmt.programs.prettier.settings.pluginSearchDirs = [
>       "${self'.packages.prettier-plugin-tailwindcss}/lib"
>     ];
> ```

Prettier remove the `pluginSearchDirs` option so that is probably why it's not working.

**Comment by Toby222** [comment]
How do you set plugin options? e.g. `@prettier/plugin-xml` has the option `xmlWhitespaceSensitivity`, but you can't set it in settings because it does not exist.

**Comment by jukremer** [comment]
Finally got it working, had to update the plugin which is in nixpkgs. 

```nix
{ pkgs, ... }:
{
  projectRootFile = "flake.nix";
  programs = {
    nixfmt.enable = true;
    gofmt.enable = true;
    taplo.enable = true;
    prettier = {
      enable = true;
      settings = {
        plugins = [
          "${pkgs.prettier-plugin-go-template}/lib/node_modules/prettier-plugin-go-template/lib/index.js"
        ];
        overrides = [
          {
            files = [ "*.html" ];
            options.parser = "go-template";
          }
        ];
      };
    };
  };
  settings.global.excludes = [
    "public/**"
    "static/**"
  ];
}
```

**Comment by GGORG0** [comment]
Is it, in any way, possible to use a Prettier plugin that's not in nixpkgs, while also at the same time using Yarn Berry (with PnP)?
In my case, if I use this `.prettierrc.cjs`:
```js
console.log(require.resolve('prettier-plugin-tailwindcss'));

module.exports = {
  plugins: [require.resolve('prettier-plugin-tailwindcss')],
};
```
It works when running Prettier normally, but not inside Nix (because both node_modules and the .yarn's cache are gitignored).
The path printed by the console.log is this very ugly path:
```
/home/ggorg/Documents/Programming/<project>/.yarn/__virtual__/prettier-plugin-tailwindcss-virtual-17a95f4a43/4/.yarn/berry/cache/prettier-plugin-tailwindcss-npm-0.6.9-b8379758a4-10c0.zip/node_modules/prettier-plugin-tailwindcss/dist/index.mjs
```

**Comment by jukremer** [comment]
I think this approach only works for packages that are in nixpkgs.

**Comment by silky** [comment]
@jukremer why? I defined the package myself above; so it should be no different.
